File: src/main/scala/com/tardis/common/TardisManager.java
package com.tardis.common;

import com.tardis.common.dimensions.InyardSavedData;
import com.temportalist.origin.library.common.lib.LogHelper;
import com.temportalist.origin.library.common.lib.vec.V3O;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.server.MinecraftServer;
import net.minecraft.util.EnumFacing;
import net.minecraft.world.WorldSavedData;
import net.minecraft.world.WorldServer;
import net.minecraftforge.common.DimensionManager;
import net.minecraftforge.fml.common.FMLCommonHandler;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 * This is the class that will handle interactions between the following: ~ Tardis -> Dimension (interior) ~ Console -> Dimension ~ Dimension -> Tardis
 *
 * @author TheTemportalist XCompWiz
 */
public class TardisManager {

	public static final List<Integer> registeredDims = new ArrayList<Integer>();
	private static final HashMap<Integer, EntityTardis> tardi = new HashMap<Integer, EntityTardis>();
	private static final HashMap<Integer, String> dimNames = new HashMap<Integer, String>();
	private static final HashMap<Integer, V3O> doors = new HashMap<Integer, V3O>();

	public static final int providerID = 1210950780;

	/*
	public static void registerConsole(World world, boolean place) {
		int dim = world.provider.getDimensionId();
		if (TardisManager.tardi.containsKey(dim)) {
			if (place) {
				TardisManager.consoles.add(dim);
			}
			else {
				TardisManager.consoles.remove(TardisManager.consoles.indexOf(dim));
			}
		}
	}
	*/

	public static void registerDimensions(boolean isRegistering) {
		for (Integer id : TardisManager.registeredDims) {
			if (isRegistering) DimensionManager.registerDimension(id, TardisManager.providerID);
			else DimensionManager.unregisterDimension(id);
		}
	}

	public static void openInterface(EntityPlayer player) {
		PlayerTardis.open(TardisManager.getTardisForDimension(
				player.getEntityWorld().provider.getDimensionId()
		), player);
	}

	public static void registerTardis(EntityTardis tardis) {
		if (FMLCommonHandler.instance().getMinecraftServerInstance() == null)
			throw new RuntimeException("Cannot create dimension client-side.");
		int id = DimensionManager.getNextFreeDimId();
		if (TardisManager.tardi.containsKey(id))
			LogHelper.info(Tardis.MODID(), "ERROR: id " + id + " already exists");
		else {
			TardisManager.registeredDims.add(id);
			DimensionManager.registerDimension(id, TardisManager.providerID);
			tardis.setInteriorDimension(id);
			TardisManager.tardi.put(id, tardis);
			TardisManager.dimNames.put(id, "TARDISDim" + id);
			System.out.println("Registered: " + TardisManager.dimNames.get(id));
			TardisManager.doors.put(id, V3O.ZERO());

			TardisManager.setData(id).markDirty();

		}
	}

	//TODO finish this method based on starting point in dimension

	/**
	 * Moves player through the passed tardis' door (transfer player to tardis interior)
	 *
	 * @param player The player passing into the interior
	 * @param tardis The tardis that is the transfer unit
	 * @param into   True if going into interior, false if coming out
	 */
	public static void movePlayerThroughDoor(EntityPlayer player, EntityTardis tardis,
			boolean into) {
		V3O pos;
		int dim;
		if (into) {
			dim = tardis.getInteriorDimension();

			WorldSavedData data = TardisManager.load(dim);
			System.out.println(TardisManager.doors.toString());

			//WorldServer dimWorld = DimensionManager.getWorld(dim);
			V3O doorPos = TardisManager.doors.get(dim); // todo this isnt saved and therefore cleared on load
			//IBlockState doorState = dimWorld.getBlockState(doorPos.toBlockPos());
			//EnumFacing facing = (EnumFacing)doorState.getValue(BlockDoor.FACING);
			if (doorPos == null) return;
			pos = doorPos.$plus(EnumFacing.NORTH);

		}
		else {
			dim = tardis.getEntityWorld().provider.getDimensionId();
			pos = new V3O(tardis);
			// todo translate pos based on tardis rotation (where tardis doors are frontwards)
		}
		// todo change to encapsulated method in Teleport
		/*
		if (player instanceof EntityPlayerMP && player.getEntityWorld() instanceof WorldServer) {
			((EntityPlayerMP)player).mcServer.getConfigurationManager().transferPlayerToDimension(
					(EntityPlayerMP)player, dim,
					new TeleporterCore((WorldServer) player.getEntityWorld())
			);
		}
		Teleport.toPoint(player, pos);
		*/
		//Teleport.toDimensionPoint(player, pos, dim);

	}

	public static String getDimName(int dimid) {
		return TardisManager.dimNames.get(dimid);
	}

	public static void setDimName(int dimid, String name) {
		TardisManager.dimNames.put(dimid, name);
	}

	public static EntityTardis getTardisForDimension(int dimid) {
		return TardisManager.tardi.containsKey(dimid) ? TardisManager.tardi.get(dimid) : null;
	}

	public static void setTardis(int dimid, EntityTardis tardis) {
		TardisManager.tardi.put(dimid, tardis);
	}

	public static V3O getDoorPos(int dimid) {
		return TardisManager.doors.get(dimid);
	}

	public static void setDoorPos(int dimid, V3O pos) {
		TardisManager.doors.put(dimid, pos);
	}

	public static WorldServer getWorld(int dimid) {
		return MinecraftServer.getServer().worldServerForDimension(dimid);
	}

	public static WorldSavedData setData(int dimid) {
		String key = TardisManager.getDimName(dimid);
		InyardSavedData data = new InyardSavedData(key);
		data.setDim(dimid);
		TardisManager.getWorld(dimid).getMapStorage().setData(key, data);
		return data;
	}

	public static WorldSavedData load(int dimid) {
		String name = TardisManager.getDimName(dimid);
		WorldServer world = TardisManager.getWorld(dimid);
		WorldSavedData data = world.getMapStorage().loadData(
				InyardSavedData.class, name
		);
		if (data == null) {
			System.out.println("Data was null");
			data = new InyardSavedData(name);
			world.getMapStorage().setData(name, data);
		}
		return data;
	}

	/*
	public static boolean hasConsole(int dimid) {
		return TardisManager.tardi.containsKey(dimid) && TardisManager.consoles.contains(dimid);
	}
	*/

}

File: src/main/scala/com/tardis/common/dimensions/TardisManager.scala
package com.tardis.common.dimensions

import java.io.File
import java.util
import java.util.List

import com.tardis.common.EntityTardis
import com.temportalist.origin.api.common.lib.vec.V3O
import com.temportalist.origin.api.common.utility.{WorldHelper, Teleport, Scala}
import com.temportalist.origin.library.common.utility.WorldHelper
import net.minecraft.entity.player.{EntityPlayerMP, EntityPlayer}
import net.minecraft.server.MinecraftServer
import net.minecraft.util.MathHelper
import net.minecraft.world.storage.MapStorage
import net.minecraft.world.{World, WorldProvider}
import net.minecraftforge.common.DimensionManager
import net.minecraftforge.common.util.ForgeDirection

/**
 *
 *
 * @author TheTemportalist
 */
object TardisManager {

	private final val filePrefix: String = "TardisDim"
	final var registeredDims: util.List[Int] = null
	// todo config option
	final var providerID: Int = 543165347
	private final val providerClass: Class[_ <: WorldProvider] = classOf[WorldProviderInyard]

	def registerProviderType(): Unit = {
		DimensionManager.registerProviderType(this.providerID, this.providerClass, true)
	}

	def getDimensionList(dir: File): List[Int] = {
		val dims: List[Int] = new util.ArrayList[Int]()
		for (file: File <- dir.listFiles()) {
			if (file.getName.startsWith(this.filePrefix) && file.getName.endsWith(".dat")) {
				try {
					val fileName: String = file.getName
					dims.add(Integer.parseInt(fileName.substring(
						filePrefix.length, fileName.length - 4
					)))
				}
				catch {
					case e: Exception => e.printStackTrace()
				}
			}
		}
		dims
	}

	def registerDimsFromDir(dir: File): Unit = {
		this.registeredDims = this.getDimensionList(dir)
		Scala.foreach(this.registeredDims, (index: Int, dimid: Int) => {
			if (!DimensionManager.isDimensionRegistered(dimid)) {
				DimensionManager.registerDimension(dimid, this.providerID)
				println("TardisManager: registered dim " + dimid)
			}
		})
	}

	def unregisterDims(): Unit = {
		Scala.foreach(this.registeredDims, (index: Int, dimid: Int) => {
			DimensionManager.unregisterDimension(dimid)
		})
		this.registeredDims = null
	}

	def getDimName(dimID: Int): String = this.filePrefix + dimID

	def getDimData(world: World): InyardData = this.getDimData(
		world.provider.dimensionId,
		!world.isRemote
	)

	def getDimData(dimID: Int, isServer: Boolean): InyardData = {
		if (DimensionManager.getProviderType(dimID) == this.providerID) {
			val dimName: String = this.getDimName(dimID)
			val storage: MapStorage = this.getStorage(isServer)
			if (storage != null) {
				var data: InyardData = storage.loadData(
					classOf[InyardData], dimName
				).asInstanceOf[InyardData]
				if (data == null) {
					data = new InyardData(dimName)
					storage.setData(dimName, data)

					data.setDim(dimID)
					data.setDoorPos(V3O.ZERO + V3O.UP)

					data.markDirty()
				}
				return data
			}
			else {
				println("Storage was null on " + (if (isServer) "server" else "client"))
			}
		}
		else {
			println("[Tardis] Dim " + dimID + " not of tardis provider type " + this.providerID)
		}
		null
	}

	def getStorage(isServer: Boolean): MapStorage = {
		/*
		if (isServer)
		else {
			println("Need to create a client storage, returning null for client side storage fetch")
			null
		}
		*/
		DimensionManager.getWorld(0).perWorldStorage
	}

	def createDimension(): InyardData = {
		val dimID: Int = DimensionManager.getNextFreeDimId
		if (WorldHelper.isClient())
			throw new RuntimeException("Cannot create dimensions client-side")
		this.registeredDims.add(dimID)
		DimensionManager.registerDimension(dimID, this.providerID)
		//PacketHandler.sendToClients(Tardis.MODID, new PacketDimensionRegistration(dimID))
		this.getDimData(dimID, isServer = false)
	}

	def registerTardis(tardis: EntityTardis): Unit = {
		val data: InyardData = this.createDimension()
		if (data != null) {
			//println ("registered tardis @ " + tardis.worldObj.provider.getDimensionId + " | " + tardis.getUniqueID)
			tardis.setInteriorDimension(data.getDim())
			data.setTardis(tardis)
			data.markDirty()
		}
		else {
			println("Data was null and could not assign tardis details")
		}
	}

	def getTardis(world: World): EntityTardis = this.getTardis(
		world.provider.dimensionId,
		!world.isRemote
	)

	def getTardis(dimID: Int, isServer: Boolean): EntityTardis = {
		this.getDimData(dimID, isServer).getTardis()
	}

	def movePlayerIntoTardis(player: EntityPlayer, tardis: EntityTardis): Unit = {
		val dimID: Int = tardis.getInteriorDimension()
		val data: InyardData = this.getDimData(dimID, player.getEntityWorld.isRemote)
		if (data == null) {
			println("null data")
			return
		}

		val spawn: V3O = data.getSpawnPoint(0) + V3O.CENTER.suppressedYAxis()
		//Teleport.toDimensionPoint(player, spawn, dimID)

		if (WorldHelper.isServer() && !player.isDead) {
			println("teleport")
			MinecraftServer.getServer.getConfigurationManager.transferPlayerToDimension(
			player.asInstanceOf[EntityPlayerMP], dimID, new TeleporterCo
			)
		}

	}

	def movePlayerOutOfTardis(player: EntityPlayer): Unit = {
		val tardis: EntityTardis = this.getTardis(player.getEntityWorld)
		if (tardis == null) {
			println("null tardis")
			return
		}

		// todo use that helper function mentioned in ItemPlacer
		val facing: Int = MathHelper.floor_double(((tardis.rotationYaw * 4F) / 360F) + 0.5D) & 3
		val convertedFacing =
			facing match {
				case 0 => 3
				case 1 => 4
				case 2 => 2
				case 3 => 5
				case _ => 2
			}
		val direction: V3O = new V3O(ForgeDirection.getOrientation(convertedFacing)) * 1.5
		val point: V3O = new V3O(tardis) + direction
		Teleport.toDimension(player, tardis.worldObj.provider.dimensionId)
		Teleport.toPoint(player, point)
	}

}
